### ISSUE TITLE
Make sh_binary deps and data do more predictable things.

### ISSUE THREAD

**Issue by nmpowell** [thought-machine/please#2434]
Example def:
```
sh_binary(
    name = "script",
    data = [
        "//path/to/some/go:binary_ONE",
    ],
    main = "script.sh",
    visibility = ["PUBLIC"],
    deps = [
        "//common/bash:common_library",
        "//path/to/some/other:binary_TWO",
    ],
)
```
If we call binary_TWO inside `script.sh` and happen to run this target in parallel, we seem to get messages like:
```
plz-out/bin/path/to/the/script.sh: line 82: /var/tmp/work/plz-out/bin/path/to/the/script/path/to/some/go/binary_TWO: Permission denied
```
which suggests binary_TWO might be being overwritten or at least having its permissions reset when the sh_binary extracts.

A possible fix is to instead use the sh_binary's `data` to include `binary_TWO`, which might be more stable and not rebuild / re-extract each time.

The `deps` are useful, though, because their output path is predictable for running in the script itself (see e.g. for the common_library above).

So it would be nice to have both:
- predictable paths for `data` which the `script.sh` can use, rather than needing to hardcode `plz-out/bin/the/known/path/to/the/data`
- a write lock or something for deps which might prevent parallel calls to deps that are being unzipped (?)

Hope that makes sense.

**Comment by nmpowell** [comment]
Additionally for the `deps` on this target, if we leave `common_library` as a `dep`, it will be zipped up into the binary and unzipped to a subdirectory when run. This is great, because the path is predictable. But with Please's standard sh_binary behaviour of unzipping, `[unzip -qo $0 -d $(dirname $0)](https://github.com/thought-machine/please/blob/0d431f21c89db7cfc1034a6b0183c492b344d564/rules/sh_rules.build_defs#L8)`, if this target is called in parallel e.g. with a `plz run parallel`, there seems to be some clobbering of those output `deps` as they are repeatedly re-extracted over one another, and we get messages like:
```
error:  cannot delete old plz-out/bin/cloud/accounts/auth/wrapper/common/bash/util_lib.sh
No such file or directory
```
and the `source`d dependency script does not appear to be found, either, giving messages like `command not found` for imported functions from the sourced script.
